src/Orleans.BroadcastChannel/BroadcastChannelConsumerExtension.cs

```diff
@@ -0,0 +1,99 @@
+using System;
+using System.Collections.Concurrent;
+using System.Threading.Tasks;
+using Orleans.Runtime;
+
+namespace Orleans.BroadcastChannel
+{
+    internal interface IBroadcastChannelConsumerExtension : IGrainExtension
+    {
+        Task OnError(InternalChannelId streamId, Exception exception);
+        Task OnPublished(InternalChannelId streamId, object item);
+    }
+
+    internal class BroadcastChannelConsumerExtension : IBroadcastChannelConsumerExtension
+    {
+        private readonly ConcurrentDictionary<InternalChannelId, ICallback> _handlers = new();
+        private readonly IOnBroadcastChannelSubscribed _subscriptionObserver;
+        private AsyncLock _lock = new AsyncLock();
+
+        private interface ICallback
+        {
+            Task OnError(Exception exception);
+
+            Task OnPublished(object item);
+        }
+
+        private class Callback<T> : ICallback
+        {
+            private readonly Func<T, Task> _onPublished;
+            private readonly Func<Exception, Task> _onError;
+
+            private static Task NoOp(Exception _) => Task.CompletedTask;
+
+            public Callback(Func<T, Task> onPublished, Func<Exception, Task> onError)
+            {
+                _onPublished = onPublished;
+                _onError = onError ?? NoOp;
+            }
+
+            public Task OnError(Exception exception) => _onError(exception);
+
+            public Task OnPublished(object item)
+            {
+                return item is T typedItem
+                    ? _onPublished(typedItem)
+                    : _onError(new InvalidCastException($"Received an item of type {item.GetType().Name}, expected {typeof(T).FullName}"));
+            }
+        }
+
+        public BroadcastChannelConsumerExtension(IGrainContextAccessor grainContextAccessor)
+        {
+            _subscriptionObserver = grainContextAccessor.GrainContext?.GrainInstance as IOnBroadcastChannelSubscribed;
```

Comment:
It seems like this is not nullable: it's accessed unconditionally further down.
In that case, perhaps a direct cast would be more appropriate.